tool/tsh/tsh.go

```diff
@@ -3132,9 +3140,29 @@ func makeClientForProxy(cf *CLIConf, proxy string, useProfileLogin bool) (*clien
 	// addresses from the proxy (this is what Ping does).
 	if cf.IdentityFileIn != "" {
 		log.Debug("Pinging the proxy to fetch listening addresses for non-web ports.")
-		if _, err := tc.Ping(cf.Context); err != nil {
+		pr, err := tc.Ping(cf.Context)
+		if err != nil {
 			return nil, trace.Wrap(err)
 		}
+		tc.LoadAllHostCAs = pr.Auth.LoadAllHostCAs
+
+		if tc.LoadAllHostCAs {
+			sites, err := tc.LocalAgent().GetClusterNames()
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			hostAuthFunc, err := key.HostKeyCallbackForClusters(cf.InsecureSkipVerify, sites)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			tc.Config.HostKeyCallback = hostAuthFunc
+			if len(key.TLSCert) > 0 {
+				tc.Config.TLS, err = key.TeleportClientTLSConfig(nil, sites)
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+			}
+		}
```

Comment:
If the logic for whether we need to load all CAs was encapsulated to NewClient like mentioned previously, all this would probably not be needed either.

Reply:
For identity files, we don't know if we should do any of this setup until after we ping, which I don't think we can do until after we create the client.